### PR TITLE
BUG: underflow on Timestamp creation

### LINKED ISSUE #14415: Inconsistent behavior in Timestamp for large negative values
#### A small, complete example of the issue

```python
# Your code here
In [1]: import numpy as np; import pandas as pd
In [3]: pd.Timestamp(np.iinfo(np.int64).min + 80000000000000)
Out[3]: Timestamp('2262-04-11 22:26:03.145224192')

In [4]: pd.Timestamp(np.iinfo(np.int64).min + 90000000000000)
Out[4]: Timestamp('1677-09-22 01:12:43.145224192')

In [5]: pd.Timestamp(np.iinfo(np.int64).min + 85000000000000)
---------------------------------------------------------------------------
OutOfBoundsDatetime                       Traceback (most recent call last)
<ipython-input-5-47bf9ab2ad0f> in <module>()
----> 1 pd.Timestamp(np.iinfo(np.int64).min + 85000000000000)

pandas/tslib.pyx in pandas.tslib.Timestamp.__new__ (pandas/tslib.c:9932)()

pandas/tslib.pyx in pandas.tslib.convert_to_tsobject (pandas/tslib.c:26453)()

pandas/tslib.pyx in pandas.tslib._check_dts_bounds (pandas/tslib.c:30034)()

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 2262-04-11 23:49:23
```

#### Expected Output

I would expect all the values listed above to evaluate to dates in the 1600s.  The behavior here looks like there might be a signed integer overflow happening in C somewhere, which is a little nerve-wracking since that's undefined behavior unless pandas is compiled with `-fwrapv`.

#### Output of ``pd.show_versions()``

<details>
# Paste the output here
In [8]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.0-16-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.19.0
nose: None
pip: 8.1.2
setuptools: 28.3.0
Cython: None
numpy: 1.11.2
scipy: None
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by chris-b1**
- [x] closes #14415
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tslib.pyx [hunk 1]
```diff
@@ -24,6 +24,11 @@ from cpython cimport (
     PyUnicode_AsUTF8String,
 )
 
+
+cdef extern from "headers/stdint.h":
+    enum: INT64_MAX
+    enum: INT64_MIN
+
```
Comment (jreback) [inline pandas/tslib.pyx]: shouldn't these be imported from the numpy headers? for consistency

File: pandas/tslib.pyx [hunk 2]
```diff
@@ -904,10 +909,9 @@ cpdef object get_value_box(ndarray arr, object loc):
 
 
 # Add the min and max fields at the class level
-# These are defined as magic numbers due to strange
-# wraparound behavior when using the true int64 lower boundary
-cdef int64_t _NS_LOWER_BOUND = -9223285636854775000LL
-cdef int64_t _NS_UPPER_BOUND = 9223372036854775807LL
+# INT64_MIN is reserved for NaT
+cdef int64_t _NS_LOWER_BOUND = INT64_MIN + 1
```
Comment (jreback) [inline pandas/tslib.pyx]: why change this?
iirc some platforms don't play nice with this definition

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tslib.pyx (hunk 1 above)
Comment (chris-b1) [inline pandas/tslib.pyx]: This is how we do it in `lib.pyx`, but agreed it probably makes more sense to pull it from the numpy headers.

Comment (jreback) [inline pandas/tslib.pyx]: ok no big deal 
ideally should be consistent across the code 
maybe just import into util.pxd and use from there?

File: pandas/tslib.pyx (hunk 2 above)
Comment (jreback) [inline pandas/tslib.pyx]: it's fine
i don't know why these were hard coded in the first place

Comment (chris-b1) [inline pandas/tslib.pyx]: On further look, I'll change the lower bound back - in our compat with python datetime we assume in several places that the min value has a 0 nanosecond unit - I think it _could_ be worked around, but fairly invasive for another 800 ns of span.

**Comment by jreback** [comment]
lgtm. Push again though (you might need to ``git commmit --all --amend -C amend`` to generate a new has. Want to make sure that appveyor is working. I just turned it back on (was set to old pydata/pandas repo)

**Comment by jreback** [comment]
lgtm. rebase and merge.

**Pull request merged**